include/deal.II/base/symmetric_tensor.h

```diff
@@ -3521,7 +3521,216 @@ outer_product(const SymmetricTensor<2, dim, Number> &t1,
   return tmp;
 }
 
+/**
+ * Perform a spectrum decomposition of a 2nd-order symmetric tensor \a
+ * original_tensor given as the input argument, \f[ \mathrm{original\_tensor} =
+ * \sum_i \lambda_i \, \boldsymbol{n}_i \otimes \boldsymbol{n}_i, \f] where
+ * $\lambda_i$ is the eigenvalue, and $\boldsymbol{n}_i$ is the corresponding
+ * eigenvector. The outputs are the positive part \a positive_part_tensor and
+ * negative part \a negative_part_tensor of the input tensor,
+ * that is,
+ * \f[
+ *   \mathrm{positive\_part\_tensor} = \sum_i <\lambda_i>_+ \boldsymbol{n}_i
+ * \otimes \boldsymbol{n}_i, \quad \mathrm{negative\_part\_tensor} = \sum_i
+ * <\lambda_i>_- \boldsymbol{n}_i \otimes \boldsymbol{n}_i, \f] where
+ * $<\lambda_i>_+ = \mathrm{max}\{ \lambda_i, 0 \}$ and
+ * $<\lambda_i>_- = \mathrm{min}\{ \lambda_i, 0 \}$. Obviously,
+ * \f[
+ *   \mathrm{positive\_part\_tensor}  + \mathrm{negative\_part\_tensor} =
+ * \mathrm{original\_tensor}. \f]
+ *
+ * @param[in] original_tensor The 2nd-order symmetric tensor to be split into
+ * the positive and negative parts
+ * @param[out] positive_part_tensor The positive part of the input tensor in
+ * which the eigenvalues are positive or zero
+ * @param[out] negative_part_tensor The negative part of the input tensor in
+ * which the eigenvalues are negative or zero
+ *
+ * @relatesalso SymmetricTensor
+ */
+template <int dim, typename Number>
+void
+positive_negative_split(const SymmetricTensor<2, dim, Number> &original_tensor,
+                        SymmetricTensor<2, dim, Number> &positive_part_tensor,
+                        SymmetricTensor<2, dim, Number> &negative_part_tensor)
+{
+  Assert(dim <= 3, ExcMessage("dim should not be larger than 3."));
+
+  std::array<std::pair<Number, Tensor<1, dim, Number>>, dim> eigen_system;
+  std::vector<Number>                                        eigen_values(dim);
+  std::vector<Tensor<1, dim, Number>>                        eigen_vectors(dim);
+
+  eigen_system = eigenvectors(original_tensor);
+
+  for (int i = 0; i < dim; i++)
+    {
+      eigen_values[i]  = eigen_system[i].first;
+      eigen_vectors[i] = eigen_system[i].second;
+    }
+
+  positive_part_tensor = 0;
+  for (int i = 0; i < dim; i++)
+    positive_part_tensor +=
+      std::fmax(eigen_values[i], 0.0) *
+      symmetrize(outer_product(eigen_vectors[i], eigen_vectors[i]));
```

Comment:
These are unnecessary but heavy computations if you're going to multiply things by zero. This may be better written as follows:
```suggestion
    if (eigen_values[i] > 0)
      positive_part_tensor +=
        eigen_values[i] *
        symmetrize(outer_product(eigen_vectors[i], eigen_vectors[i]));
```